src/util/expr_iterator.h

```diff
@@ -0,0 +1,239 @@
+  depth_iterator_t &operator++()
+  {
+    PRECONDITION(!m_stack.empty());
+    while(true)
+    {
+      if(m_stack.back().it==m_stack.back().end)
+      {
+        m_stack.pop_back();
+        if(m_stack.empty())
+          break;
+      }
+      // Check eg. if we haven't seen this node before
+      else if(this->downcast().push_expr(*m_stack.back().it))
+        break;
+      m_stack.back().it++;
+    }
+    return this->downcast();
+  }
+
+  /// Post-increment operator
+  /// Expensive copy. Avoid if possible
+  depth_iterator_t operator++(int)
+  {
+    depth_iterator_t tmp(*this);
+    this->operator++();
+    return tmp;
+  }
+
+  /// Dereference operator
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt &operator*() const
+  {
+    PRECONDITION(!m_stack.empty());
+    return m_stack.back().expr.get();
+  }
+
+  /// Dereference operator (member access)
+  /// Dereferencing end() iterator is undefined behaviour
+  const exprt *operator->() const
+  { return &**this; }
+
+protected:
+  /// Create end iterator
+  depth_iterator_baset()=default;
+
+  /// Create begin iterator, starting at the supplied node
+  explicit depth_iterator_baset(const exprt &root)
+  { this->push_expr(root); }
+
+  /// Destructor
+  /// Protected to discourage upcasting
+  ~depth_iterator_baset()=default;
+
+  depth_iterator_baset(const depth_iterator_baset &)=default;
+  depth_iterator_baset(depth_iterator_baset &&other)
+  { m_stack=std::move(other.m_stack); }
+  depth_iterator_baset &operator=(const depth_iterator_baset&)=default;
+  depth_iterator_baset &operator=(depth_iterator_baset &&other)
+  { m_stack=std::move(other.m_stack); }
+
+  /// Obtain non-const exprt reference. Performs a copy-on-write on
+  /// the root node as a side effect. To be called only if a the root
+  /// is a non-const exprt. Do not call on the end() iterator
+  exprt &mutate()
+  {
+    PRECONDITION(!m_stack.empty());
+    auto expr=std::ref(const_cast<exprt &>(m_stack.front().expr.get()));
+    for(auto &state : m_stack)
+    {
+      auto &operands=expr.get().operands();
+      const auto i=operands.size()-(state.end-state.it);
+      const auto it=operands.begin()+i;
+      state.expr=expr.get();
+      state.it=it;
+      state.end=operands.end();
+      if(!(state==m_stack.back()))
+      {
+        expr=std::ref(*it);
+      }
+    }
+    return expr.get();
+  }
+
+  /// Pushes expression onto the stack
+  /// If overridden, this function should be called from the inheriting
+  /// class by the override function
+  /// \return true if element was successfully pushed onto the stack,
+  /// false otherwise
+  /// If returning false, child will not be iterated over
+  bool push_expr(const exprt &expr)
+  {
+    m_stack.emplace_back(expr, expr.operands().begin(), expr.operands().end());
+    return true;
+  }
+
+private:
+  typedef typename std::vector<exprt>::const_iterator operands_iteratort;
+
+  struct expr_statet final
```

Comment:
Sorry, just noticed this: It seems that `expr_statet` isn't dependent on `depth_iterator_t` and therefore I don't think it should be a member type. At the moment the compiler will create a separate struct for each instantiation with a different `depth_iterator_t` which is probably not what we want.